### ISSUE TITLE
Drop the *.rule format and only use the standard *.rules

### ISSUE THREAD

**Issue by sed-i** [canonical/prometheus-operator#153]
From looking at an internal client's rule files:
1. all the rules are in the same folder (not divided into group dirs)
1. in all rule files there is only one group per file, and group name matches filename
1. in all rule files there are multiple alerts per file

This is contrast to our current working assumptions:
1. Only one rule per file, and the "groups" level is omitted.
1. All rules from files in the same directory are loaded into a single group.
1. Rule files are loaded recursively and the relpath to root becomes part of the group name.

Seems like prom lib should support "proper" rules files so that users do not need to refactor their entire rules collection.
Should we drop the *.rule format we rolled on our own and only focus on the standard *.rules?

**Comment by sed-i** [comment]
>     1. We anyhow cannot keep the group names unmodified because of name clashes if one deploys the same charm multiple times

afaiu, name clashes are not possible because topology identifiers are part of the group name

>     2. One group per embedded alert rules from one deployed charm makes those rules easier to find in Prometheus.

If charm author provides their own group name (to which topology identifiers could be appended) then finding them in prom is even easier.

https://github.com/canonical/prometheus-operator/blob/cda2e753f8cbad7c81269fcb043844ede8002fc5/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L604-L610

> There was no intention to support recursive rule folders.

There is now :)

> All rules read from a client charm are placed into the same *.rules files.

That is still the case, even with nested rules dirs:

https://github.com/canonical/prometheus-operator/blob/cda2e753f8cbad7c81269fcb043844ede8002fc5/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L612-L615

**Comment by sed-i** [comment]
Another reason to keep one standard is recording rules. Here's a pastie from loki but I think it's equivalent for prom.

If we take a normal, standard rules file
```yaml
groups:
  - name: should_fire
    rules:
      - alert: HighPercentageError
        expr: |
          sum(rate({app="foo", env="production"} |= "error" [5m])) by (job)
            /
          sum(rate({app="foo", env="production"}[5m])) by (job)
            > 0.05
        for: 10m
        labels:
            severity: page
        annotations:
            summary: High request latency
```
and dumb it down, we get:
```yaml
alert: HighPercentageError_topo
expr: |
  sum(rate({app="foo", env="production", %%juju_topology%%} |= "error" [5m])) by (job)
    /
  sum(rate({app="foo", env="production", %%juju_topology%%}[5m])) by (job)
    > 0.05
for: 10m
labels:
    severity: page
annotations:
    summary: High request latency
```
Now let's see what happens with recording rules:
```yaml
name: NginxRules
interval: 1m
rules:
  - record: nginx:requests:rate1m
    expr: |
      sum(
        rate({container="nginx"}[1m])
      )
    labels:
      cluster: "us-central1"
```
dumb it down:
```yaml
record: nginx:requests:rate1m
expr: |
  sum(
    rate({container="nginx", %%juju_topology%%}[1m])
  )
labels:
  cluster: "us-central1"
```
and we lost `interval`.

**Comment by mmanciop** [comment]
I was not aware that groups can override the evaluation interval. I do not see this as a “we must henceforth drop single rules!”, but it is a good finding. What else can groups do?

**Comment by sed-i** [comment]
Closing this as there is now a PR for supporting both formats
https://github.com/canonical/prometheus-operator/pull/161

**Comment by balbirthomas** [comment]
> Closing this as there is now a PR for supporting both formats #161

Reopening this ticket for review. Adding a Closes clause in the PR should automatically close this ticket when the PR is merged.

**Comment by balbirthomas** [comment]
Could you please clarify :

1. Is it the requirement that recording and alerting rules should be combined into the same charm folder and use the same code flow for processing both ? Note the above discussion is making that case for supporting both "*.rule" and "*.rules" by  mixing examples of both recording and alerting rules. At present the default folder name for "alert rules"  is "prometheus_alert_rules" !  More importantly as far I can tell from the Prometheus [alerting rules docs](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) there are *no groups specific parameters*  for alerting rules (such as shown in the example above for recording rules). The only thing an alerting rule group has a "group name" and a "list of rules" or am I missing something here ?

2. Group names using Juju topology are fairly unambiguous. What is the use case for one charm submitting multiple alert rule groups (the only difference between them being a different name) ?

3. In fact the Prometheus alert rules file can contain multiple groups in a single file. If the desire is to support the "full prometheus format" . Shouldn't we be supporting multiple alert rule groups in a single file ? Or is there some reason not to do this ? Do charms need to support the "full Prometheus format" or can we effect some simplification to the benefit of the system admins and charm writers ?

3. What is the use case for recursive rules folders ? 

PS: Perhaps the title of this ticket needs to be changed.

**Comment by sed-i** [comment]
This issue is for supporting only one format, and that proposal was rejected.
Your points are valid but let's continue this discussion under https://github.com/canonical/prometheus-operator/pull/161

**Comment by simskij** [comment]
> This issue is for supporting only one format, and that proposal was rejected.
> Your points are valid but let's continue this discussion under #161

With that in mind, closing this again.
